### PR TITLE
Adding circuit executor classes and shot-branching

### PR DESCRIPTION
**Pull request by doichanj**
### Summary
This PR restructures parallel simulation classes that were implemented in `StateChunk` class.
Instead of `StateChunk` this PR introduces `CircuitExecutor` classes outside of `State` classes.  

This PR also implements shot-branching revised from PR #1606 
Now the implementation is simplified in `CircuitExecutor::MultiStateExecutor` class

### Details and comments

image

### CODE REVIEW

File: src/controllers/aer_controller.hpp [hunk 1]
```diff
@@ -917,18 +565,7 @@ Result Controller::execute(std::vector<Circuit> &circuits,
```
Comment (hhorii) [inline src/controllers/aer_controller.hpp:565]: I think `Controller::execute(std::vector<Circuit> &circuits, ..)` method can be

1. Determine methods (calling `simulation_methods()`)
2. Construct `std::vector<std::shared_ptr<Executor>>`
3. Determine experiment parallelization (`set_parallelization_experiments()`) by using `Executor.required_memory_mb()`
4. Call `Executor. run_circuit()` in parallel or serial

Construction of `Executor` can be via a static method in `simulators.hpp`.

**Review by hhorii — COMMENTED** [review]
I would like to leave comments on `aer_controller.hpp` first.

File: src/simulators/circuit_executor.hpp [hunk 2]
```diff
@@ -0,0 +1,1174 @@
...
+  // results are stored independently in each process if true
+  bool accept_distributed_results_ = true;
+
+  uint_t myrank_;               // process ID
+  uint_t nprocs_;               // number of processes
+  uint_t distributed_rank_;     // process ID in communicator group
+  uint_t distributed_procs_;    // number of processes in communicator group
+  uint_t distributed_group_;    // group id of distribution
+  int_t distributed_proc_bits_; // distributed_procs_=2^distributed_proc_bits_
+                                // (if nprocs != power of 2, set -1)
```
Comment (hhorii) [inline src/simulators/circuit_executor.hpp:87]:
```
  uint_t myrank_;               // process ID
  uint_t nprocs_;               // number of processes
  uint_t distributed_rank_;     // process ID in communicator group
  uint_t distributed_procs_;    // number of processes in communicator group
  uint_t distributed_group_;    // group id of distribution
  int_t distributed_proc_bits_; // distributed_procs_=2^distributed_proc_bits_
                                // (if nprocs != power of 2, set -1)
```
Theses values are only for MPI, I think. Can we put them in `#ifdef AER_MPI`. Maybe we set some codes in `aer_controller.hpp` in it.

File: src/simulators/circuit_executor.hpp [hunk 3]
```diff
@@ -0,0 +1,1174 @@
...
+  int num_process_per_experiment_ = 1;
+
+#ifdef AER_MPI
+  // communicator group to simulate a circuit (for multi-experiments)
+  MPI_Comm distributed_comm_;
+#endif
+
+  // settings for cuStateVec
+  bool cuStateVec_enable_ = false;
```
Comment (hhorii) [inline src/simulators/circuit_executor.hpp:96]: This also can be in `#ifndef AER_THRUST_CPU` I guess.

File: src/simulators/circuit_executor.hpp [hunk 4]
```diff
@@ -0,0 +1,1174 @@
...
+    // set configuration
+    method_ = method;
+    sim_device_ = device;
+
+    set_config(config);
+    set_parallelization(circ, noise);
+
+    // Rng engine (this one is used to add noise on circuit)
+    RngEngine rng;
+    rng.set_seed(circ.seed);
```
Comment (hhorii) [inline src/simulators/circuit_executor.hpp:489]: `rng.set_seed(circ.seed);` is called in later (`run_circuit_with_sampling` or `run_circuit_shots `). It is better to use different seeds for them.

File: src/simulators/circuit_executor.hpp [hunk 5]
```diff
@@ -0,0 +1,1174 @@
...
+    state.set_global_phase(circ.global_phase_angle);
+
+    state.set_distribution(1);
+    state.set_max_matrix_qubits(max_bits);
+
+    RngEngine rng;
+    rng.set_seed(circ.seed);
+    run_with_sampling(circ, state, result, rng, circ.shots);
+  } else {
+    // Vector to store parallel thread output data
```
Comment (hhorii) [inline src/simulators/circuit_executor.hpp:649]: I'm confused that `run_circuit_with_sampling`  needs to consider about `parallel_shots_ `. If we use `sampling`, `state` will be calculated once and then get bitstrings from the `state`.

File: src/simulators/circuit_executor.hpp [hunk 6]
```diff
@@ -0,0 +1,1174 @@
...
+      } else {
+        par_results[i].save_count_data(state.creg(), save_creg_memory_);
+      }
+    }
+    state.add_metadata(par_results[i]);
+  };
+  Utils::apply_omp_parallel_for((par_shots > 1), 0, par_shots,
+                                run_circuit_lambda);
+
+  // gather cregs on MPI processes and save to result
```
Comment (hhorii) [inline src/simulators/circuit_executor.hpp:776]: The below can be in `#AER_MPI` (`gather_creg_memory ` is also), I guess.

**Review by hhorii — COMMENTED** [review]
Comments for `circuit_executor.hpp` is here:

File: src/controllers/aer_controller.hpp (hunk 1 above)
Comment (doichanj) [reply src/controllers/aer_controller.hpp:515]: I create `Controller::make_circuit_executor` function and removed `required_memory_mb` and `validate_state` from `Controller`.

File: src/simulators/circuit_executor.hpp (hunk 2 above)
Comment (doichanj) [reply src/simulators/circuit_executor.hpp:108]: There are many depending codes to these params, so I would like to keep them available for non MPI environment. But I added `#ifdef AER_MPI `to the codes we can separate.

File: src/simulators/circuit_executor.hpp (hunk 3 above)
Comment (doichanj) [reply src/simulators/circuit_executor.hpp:118]: I added `#ifdef AER_CUSTATEVEC` to this param.

File: src/simulators/circuit_executor.hpp (hunk 4 above)
Comment (doichanj) [reply src/simulators/circuit_executor.hpp:516]: Now initial rng used for noise sampling is reused in `run_circuit_xxx` functions.
I think this is the strategy used in the older code.

File: src/simulators/circuit_executor.hpp (hunk 5 above)
Comment (doichanj) [reply src/simulators/circuit_executor.hpp]: removed unnecessary functions and codes

File: src/simulators/circuit_executor.hpp (hunk 6 above)
Comment (doichanj) [reply src/simulators/circuit_executor.hpp:759]: `#ifdef AER_MPI` inserted

File: src/simulators/multi_state_executor.hpp [hunk 7]
```diff
@@ -0,0 +1,758 @@
...
+    state_t tmp;
+    return tmp.required_memory_mb(circuit.num_qubits, circuit.ops);
+  }
+
+  uint_t get_process_by_chunk(uint_t cid);
+
+protected:
+  void set_config(const Config &config) override;
+
+  void set_distribution(uint_t num_states);
```
Comment (hhorii) [inline src/simulators/multi_state_executor.hpp:106]: Could you add comments to explain this method? Originally, I found the following comment in`state_chunk.hpp` to the method of the same name excepting its argument name `uint_t nprocs`.

> // set number of processes to be distributed

File: src/controllers/aer_controller.hpp [hunk 8]
```diff
@@ -1042,809 +636,128 @@ Result Controller::execute(std::vector<Circuit> &circuits,
...
-  }
+    return std::make_shared<
+        CircuitExecutor::Executor<ExtendedStabilizer::State>>();
+  } break;
   case Method::matrix_product_state: {
-    MatrixProductState::State state;
-    return state.required_memory_mb(circ.num_qubits, circ.ops);
-  }
+    return std::make_shared<
+        CircuitExecutor::Executor<MatrixProductState::State>>();
```
Comment (hhorii) [inline src/controllers/aer_controller.hpp:741]: Why not `MultiStateExecutor` is used for MPS and others?

**Review by hhorii — COMMENTED** [review]
Reviewed Executors. I'm wondering `MultiStateExecutor` is effective if shot-branching improves any methods.

File: src/controllers/aer_controller.hpp (hunk 8 above)
Comment (doichanj) [reply src/controllers/aer_controller.hpp:741]: Only a simulation method who supports shot-branching inherits `MultiStateExecutor` at this time.

File: src/controllers/aer_controller.hpp [hunk 9]
```diff
@@ -948,12 +535,14 @@ Result Controller::execute(std::vector<Circuit> &circuits,
     result.metadata.add(max_memory_mb_, "max_memory_mb");
     result.metadata.add(max_gpu_memory_mb_, "max_gpu_memory_mb");
 
+#ifdef AER_MPI
```
[the remark this answers is not in the document]
Comment (doichanj) [reply src/controllers/aer_controller.hpp]: merged ifdef blocks

File: src/framework/results/experiment_result.hpp [hunk 10]
```diff
@@ -217,19 +211,24 @@ void ExperimentResult::save_data_average(const ClassicalRegister &creg,
...
 void ExperimentResult::save_data_pershot(const ClassicalRegister &creg,
                                          const std::string &key, const T &datum,
-                                         OpType type, DataSubType subtype) {
+                                         OpType type, DataSubType subtype,
+                                         const uint_t nshots) {
   switch (subtype) {
   case DataSubType::single:
-    data.add_single(datum, key);
+    for (int_t i = 0; i < nshots; i++)
+      data.add_single(datum, key);
```
[the remark this answers is not in the document]
Comment (doichanj) [reply src/framework/results/experiment_result.hpp]: removed nshots and loop from `save_dat_pershot`

**Review by hhorii — APPROVED** [review]
Though I was not able to review the detail of implementation due to the size of changes, I believe that we can merge this PR and have time to refine implementation by the release of 0.13.0.

**Pull request merged**